hbase-client/src/main/java/org/apache/hadoop/hbase/util/PoolMap.java

```diff
@@ -45,177 +36,138 @@
  * </p>
  *
+ * <p>
+ * PoolMap is thread-safe. It does not remove elements automatically. Unused resources
+ * must be closed and removed explicitly.
+ * </p>
+ *
  * @param <K>
  *          the type of the key to the resource
  * @param <V>
  *          the type of the resource being pooled
  */
 @InterfaceAudience.Private
-public class PoolMap<K, V> implements Map<K, V> {
-  private PoolType poolType;
-
-  private int poolMaxSize;
-
-  private Map<K, Pool<V>> pools = new ConcurrentHashMap<>();
-
-  public PoolMap(PoolType poolType) {
-    this.poolType = poolType;
-  }
-
-  public PoolMap(PoolType poolType, int poolMaxSize) {
-    this.poolType = poolType;
-    this.poolMaxSize = poolMaxSize;
+public class PoolMap<K, V> {
+  private final Map<K, Pool<V>> pools;
+  private final PoolType poolType;
+  private final int poolMaxSize;
+
+   public PoolMap(PoolType poolType, int poolMaxSize) {
+     pools = new HashMap<>();
+     this.poolType = poolType;
+     this.poolMaxSize = poolMaxSize;
   }
 
-  @Override
-  public V get(Object key) {
-    Pool<V> pool = pools.get(key);
-    return pool != null ? pool.get() : null;
+  public V getOrCreate(K key, PoolResourceSupplier<V> supplier) throws IOException {
+     synchronized (pools) {
+       Pool<V> pool = pools.get(key);
+
+       if (pool == null) {
+         pool = createPool();
+         pools.put(key, pool);
+       }
+
+       try {
+         return pool.getOrCreate(supplier);
+       } catch (IOException | RuntimeException | Error e) {
+         if (pool.size() == 0) {
+           pools.remove(key);
+         }
+
+         throw e;
+       }
+     }
   }
+  public boolean remove(K key, V value) {
+    synchronized (pools) {
+      Pool<V> pool = pools.get(key);
 
-  @Override
-  public V put(K key, V value) {
-    Pool<V> pool = pools.get(key);
-    if (pool == null) {
-      pools.put(key, pool = createPool());
-    }
-    return pool != null ? pool.put(value) : null;
-  }
+      if (pool == null) {
+        return false;
+      }
 
-  @SuppressWarnings("unchecked")
-  @Override
-  public V remove(Object key) {
-    Pool<V> pool = pools.remove(key);
-    if (pool != null) {
-      removeValue((K) key, pool.get());
-    }
-    return null;
-  }
+      boolean removed = pool.remove(value);
 
-  public boolean removeValue(K key, V value) {
-    Pool<V> pool = pools.get(key);
-    boolean res = false;
-    if (pool != null) {
-      res = pool.remove(value);
-      if (res && pool.size() == 0) {
+      if (removed && pool.size() == 0) {
         pools.remove(key);
       }
+
+      return removed;
     }
-    return res;
   }
 
-  @Override
-  public Collection<V> values() {
-    Collection<V> values = new ArrayList<>();
-    for (Pool<V> pool : pools.values()) {
-      Collection<V> poolValues = pool.values();
-      if (poolValues != null) {
-        values.addAll(poolValues);
-      }
+  public void remove(K key) {
+    synchronized (pools) {
+      pools.remove(key);
     }
```

Comment:
Do we still need this method? I don't see it getting used in `AbstractRpcClient`?

Reply:
I removed unused methods.